### PR TITLE
Fix PN metadata leak, PN navigation to chat, and unexpected PNs from other accounts

### LINKED ISSUE #6772: Metadata leak in push notifications
# Problem

Reported by @hesterbruikman:

If a message sender has their mobile OS language settings set to a supported language (e.g. Korean or Spanish) the notification of the recipient is shown in this language, regardless of the anguage the recipient has set in their OS. E.g. Jinho sends a message to Philio while Jinho's phone is set to Korean, Philip receives a notification in Korean while his phone is set to English.

# Acceptance Criteria

Push notifications should not reveal metadata

cc @mandrigin @corpetty @rasom @PombeirP @corpetty

### PR DESCRIPTION
**Pull request by rasom**
The same PR as https://github.com/status-im/status-react/pull/6893. It has too many comments, quite hard to "navigate" even to find new builds.

[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

fixes #6772, #3488 and #2984, depends on https://github.com/status-im/status-go/pull/1297

### Summary:

[comment]: # (Summarise the problem and how the pull request solves it)

Pre-PR example payload:

```javascript
{
	"notification": {
		"title": "Status",
		"body": "You have a new message"
	},
	"data": {
		"msg": {
			"from": "0x04325367620ae20dd878dbb39f69f02c567d789dd21af8a88623dc5b529827c2812571c380a2cd8236a2851b8843d6486481166c39debf60a5d30b9099c66213e4",
			"to": "0x04601f0228b9c046ddf524f7c612093c0ba69a8b7ee37e3f9292626fd6bb50578f57e9f5672e79967fcef7803cfe28daf93342cb4cdb6eb9702eb759d3154b192b"
		}
	}
}
```

Same payload post-PR (notice versioning):
```javascript
{
	"data": {
		"msg-v2": {
			"from": "0x2cea3bd5",
			"to": "0xb1f89744",
			"id": "0x872653ad"
		}
	}
}
```
This PR:
- [x] Adds new message id field to the PN in the form of `hash:${msg-v2/id}`, creating a 1:1 link between Whisper messages and their respective PN on the receiver side 
- [x] Moves to using sender/receiver pubkey hashes instead of the pubkeys themselves (first 10 characters/4 bytes of sha3 hash)
  - [x] Key/values sent: `msg-v2/from`, `msg-v2/to`, `msg-v2/id`
- [x] Stores complete PN payload as JSON in db (`:push-notifications/stored`)
- [x] Moves from Notification API (where title/body were being sent with the PN) to Messaging API (Messaging API only sends a hidden message with key/value pairs) (#6772)
- [x] Maintains compatibility with older clients (receiving only, sent PNs will not be displayed on older clients)
- [x] Creates background task to receive FCM message and display appropriate PN
  - [x] ~~Only create PN if message with hash equal to `msg-v2/id` is not marked as seen~~
- [x] Removes unused `config/in-app-notifications-enabled?`/`IN_APP_NOTIFICATIONS_ENABLED`
- [x] ~~Removes dependency on NotifyUsers status-go API by leveraging react-native-firebase API~~ the API does not support some critical properties for Android and iOS, so we must continue with the implementation in status-go for the time being. Created a suggestion at https://boards.invertase.io/react-native-firebase/p/add-missing-properties-to-messagingremotemessage.
- [x] Fixes navigating to chat if the user is signed out and taps on notification (#3488).

### Review notes (optional):

This PR paves the way to fix issues like https://github.com/status-im/status-react/issues/3488, https://github.com/status-im/status-react/issues/3487, https://github.com/status-im/status-react/issues/2984

#### Platforms
Check that new apps can receive PNs from both new and old clients. I've tested Desktop and Android, but not iOS.
- Android (as sender/receiver)
- iOS (as sender/receiver)
- Desktop (as sender only)

See here for expected behavior from a technical perspective: https://github.com/invertase/react-native-firebase-docs/blob/master/docs/messaging/introduction.md#data-only-messages

#### Areas that maybe impacted
**Functional**
- 1-1 chats
- group chats

### Steps to test:
1. Open Status on device 1
1. Open Status on device 2
1. Add device 2 as a contact on device 1
1. Send a message from device 1 to device 2
1. Add device 1 as a contact of device 2
1. Put the app on device 2 in the background
1. Send another message from device 1 to device 2
1. Confirm that a Push Notification is displayed on device 2
1. Tap on the PN
1. Confirm that the app navigates to the appropriate chat

### Future steps

With this PR implemented, we can easily implement the following:

1. Show sender/chat room in PN if db is unlocked (https://github.com/status-im/status-react/issues/7043)
1. Group PNs by sender/chat (https://github.com/status-im/status-react/issues/7046)
1. Navigate to message matching partial IDs in mobile push notifications when PN is tapped #7044

status: ready 

<blockquote><div><strong><a href="https://boards.invertase.io/react-native-firebase/p/add-missing-properties-to-messagingremotemessage">Add missing properties to messaging.RemoteMessage | React Native Firebase | Invertase</a></strong></div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/invertase/react-native-firebase-docs">invertase/react-native-firebase-docs</a></strong></div><div>📖 React Native Firebase documentation source files - please use the website to browse the docs. - invertase/react-native-firebase-docs</div></blockquote>

### CODE REVIEW

File: src/status_im/ui/screens/navigation.cljs
Comment (pedropombeiro) [inline src/status_im/ui/screens/navigation.cljs]: I imagine this is temp code? If not, we should add more context here as to what we're actually logging.

File: src/status_im/init/core.cljs
Comment (pedropombeiro) [inline src/status_im/init/core.cljs]: Temp code?

**Review by pedropombeiro — APPROVED** [review]

File: src/status_im/ui/screens/navigation.cljs
Comment (rasom) [reply src/status_im/ui/screens/navigation.cljs]: yes i will remove it, currently i'm trying to debug app

File: src/status_im/init/core.cljs
Comment (rasom) [reply src/status_im/init/core.cljs]: yes

**Review by mandrigin — APPROVED** [review]

**Comment by asemiankevich** [comment]
for some reason i still dont receive PNs on my ios device. I am using iPhone 7 with iOS 11.4.1. I think the problem may be in this `Allow notifications` setting flag. I have turned it on and off in settings and it does not make any effect. Android device looks good to me.
@Serhy @churik please check on your side.

**Comment by churik** [comment]
@asemiankevich case when you turn off and on PN works for me as expected - so I can again receive PNs.
Device: IPhone 7
IOS: 11.4.1

**Comment by rasom** [comment]
@asemiankevich just checked on iphone X and 6s, both with 12.1.2, works as expected. From which device/app version do you send the messages?

**Comment by churik** [comment]
> @asemiankevich just checked on iphone X and 6s, both with 12.1.2, works as expected. From which device/app version do you send the messages?

@rasom 
From @asemiankevich : Ios 11.4.1 can't receive PN from Android 8

**Comment by rasom** [comment]
> From @asemiankevich : Ios 11.4.1 can't receive PN from Android 8

ok, and what's the app version on android?

**Comment by rasom** [comment]
Tried to turn on/off notifications on ios, can receive PNs after turning notifications on.

**Comment by asemiankevich** [comment]
> > From @asemiankevich : Ios 11.4.1 can't receive PN from Android 8
> 
> ok, and what's the app version on android?

@rasom it is android 8, version of the app is latest current PR build

**Comment by annadanchenko** [comment]
can also reproduce no PN if Android 9, pixel XL sends new message in 1:1 chat to IPhone 7 plus, iOS 11.0.3.

1. Iphone added pixel by scanning qr code and sent a new text message
2. Pixel added iPhone as a contact and sent a message to iPhone
3. App is on background on iPhone
4. Pixel sends new message to iPhone -> no PN shown but if switch to chat on iPhone then new message is there

**Comment by asemiankevich** [comment]
probably it is related to iOS version, because old versions are killing the app on background faster, not sure. It is all fine with iOS 12.

**Comment by rasom** [comment]
@asemiankevich @annadanchenko @PombeirP @mandrigin 
On my iphone X i can see PNs even after 1h in the background (probably even longer but i didn't test), on iphone 6s the app is suspended in 2-3 minutes. Currently, if app is suspended we can't receive PNs on ios. 

We can try to implement some handling on native side but not as part of this PR. I would suggest merging it as it is for now.

**Comment by asemiankevich** [comment]
agree

**Comment by asemiankevich** [comment]
moved back to testing, there are some other issues to be rechecked before merge

**Comment by rasom** [comment]
@asemiankevich make sure you are testing build # 10, that builds table is broken a bit

**Comment by rasom** [comment]
@PombeirP 
as i understand changes necessary on go side are currently reverted from `status-go`'s `develop`, do we need to file another PR on `status-go` side?

**Comment by churik** [comment]
@rasom I'm a bit confused. Should I start testing on this one (build # 10, commit`ae1acf3`)or wait for https://github.com/status-im/status-react/pull/7268#issuecomment-455077946 response?

**Comment by rasom** [comment]
@churik 
i think currently # 11 is the last build 

You don't need to wait, all necessary changes are already here. Question to Pedro is rather about the coordination of necessary changes on go side.

**Comment by pedropombeiro** [comment]
@rasom I'll build a PR on go side to add a new method side-by-side with `NotifyUsers` (I'll call it `SendMessageToUsers` if you agree, since it users the messaging API instead of the notifications API) and then we'll just need to adapt this PR for it.

**Comment by rasom** [comment]
@PombeirP ok

**Comment by churik** [comment]
@rasom 

**1) IOS crash is not fixed yet**

Steps described in https://github.com/status-im/status-react/pull/6893#issuecomment-453944160:

- Put app on ios device to background
- send message from another device
- close app on ios device
- tap notification
- app crashes

testing is still WIP

**Comment by rasom** [comment]
```
Jan 17 14:47:01 iPhonex-Roman StatusIm[20151] <Notice>: Exception '*** -[__NSDictionaryM setObject:forKeyedSubscript:]: key cannot be nil' was thrown while invoking complete on target RNFirebaseNotifications with params (
    "<null>",
    1
)
callstack: (
	0   CoreFoundation                      0x00000001d1160edc <redacted> + 252
	1   libobjc.A.dylib                     0x00000001d0331a40 objc_exception_throw + 56
	2   CoreFoundation                      0x00000001d10d8494 _CFArgv + 0
	3   CoreFoundation                      0x00000001d1051ca0 <redacted> + 932
	4   StatusIm                            0x0000000104fc6530 _ZN5folly6detail15str_to_integralIyEENS_8ExpectedIT_NS_14ConversionCodeEEEPNS_5RangeIPKcEE + 899356
	5   CoreFoundation                      0x00000001d1168630 <redacted> + 144
	6   CoreFoundation                      0x00000001d1046450 <redacted> + 292
	7   CoreFoundation                      0x00000001d1047034 <redacted> + 60
	8   StatusIm                            0x0000000104e80d68 StatusIm + 740712
	9   StatusIm                     <\M-b\M^@\M-&>
Jan 17 14:47:01 iPhonex-Roman StatusIm(CoreFoundation)[20151] <Notice>: *** Terminating app due to uncaught exception 'RCTFatalException: Exception '*** -[__NSDictionaryM setObject:forKeyedSubscript:]: key cannot be nil' was thrown while invoking complete on target RNFirebaseNotifications with params (
    "<null>",
    1
)
callstack: (
	0   CoreFoundation                      0x00000001d1160edc <redacted> + 252
	1   libobjc.A.dylib                     0x00000001d0331a40 objc_exception_throw + 56
	2   CoreFoundation                      0x00000001d10d8494 _CFArgv + 0
	3   CoreFoundation                      0x00000001d1051ca0 <redacted> + 932
	4   StatusIm                            0x0000000104fc6530 _ZN5folly6detail15str_to_integralIyEENS_8ExpectedIT_NS_14ConversionCodeEEEPNS_5RangeIPKcEE + 899356
	5   CoreFoundation                      0x00000001d1168630 <redacted> + 144
	6   CoreFoundation                      0x00000001d1046450 <redacted> + 292
	7   CoreFoundation                      0x00000001d1047034 <redacted> + 60
	8   StatusIm                            0x000<\M-b\M^@\M-&>
```

**Comment by rasom** [comment]
https://github.com/invertase/react-native-firebase/issues/1576#issuecomment-430750112 might be related...

**Comment by rasom** [comment]
@churik ios issue is fixed in # 22

**Comment by churik** [comment]
@rasom can it affect all functionality or recheck only specific issue will be enough?

**Comment by rasom** [comment]
this specific issue will be enough

**Comment by churik** [comment]
@rasom on latest build I started receive messages when app is closed (on IOS). IS it OK from the security perspective (taking into account https://github.com/status-im/status-react/pull/6893#issuecomment-453983432)?

**Comment by rasom** [comment]
This may happen only if you send messages from another version of the application, which doesn't have current changes.

**Comment by churik** [comment]
Rechecked https://github.com/status-im/status-react/pull/7268#issuecomment-455144753 on build #22 - not reproducible
Full set of performed tests is [here](https://docs.google.com/spreadsheets/d/1FYvnT7FDSHSl4bzK8lwEtd8HiIWIu4qwa38Z2Jh64ZM/edit?usp=sharing)

Want to summarise:
1) There is one **breaking change** - so users with previous builds (0.9.32 and earlier) can't receive PNs from users with last build

2) IOS users will not receive PNs when app is closed- discussion [here](https://github.com/status-im/status-react/pull/6893#issuecomment-453983432)

3) In some cases users can't receive PNs. 
Comments: https://github.com/status-im/status-react/pull/7268#issuecomment-454714949 , https://github.com/status-im/status-react/pull/7268#issuecomment-454673662

Please tell me if some of those should be reported as separate issues and fixed.

cc @rachelhamlin @annadanchenko

**Comment by mandrigin** [comment]
@rasom please squash and merge these changes 🎉

**Comment by rasom** [comment]
@mandrigin @PombeirP 
this PR contains upgraded status-go, are we good to merge or need to wait till #7229 will be merged? Also it seems that https://github.com/status-im/status-go/pull/1352 is not merged yet

**Comment by mandrigin** [comment]
@rasom but this PR doesn't use this new method, I know it was reverted for #7229, but maybe we can merge this one now, then revert the revert and update status-go in #7229. 
I believe there are some things left to be fixed on the Go side anyway.

**Comment by rasom** [comment]
ok, merged

**Comment by pedropombeiro** [comment]
Thanks for all the help @rasom! This was a tough one!

**Comment by rachelhamlin** [comment]
@churik thanks :)

We'll call out the breaking change in the next release notes. 

I think we should capture the issues with iOS and discuss priority in next core meeting.

**Pull request merged**
